L1Trigger/L1TTrackMatch/plugins/DisplacedVertexProducer.cc

```diff
@@ -0,0 +1,481 @@
+  std::vector<std::pair<Track_Parameters, edm::Ptr<TrackingParticle>>> selectedTracksEmulationWithTruth;
+
+  //Simulation track selection loop
+  for (iterL1Track = TTTrackHandle->begin(); iterL1Track != TTTrackHandle->end(); iterL1Track++) {
+    edm::Ptr<TTTrack<Ref_Phase2TrackerDigi_>> l1track_ptr(TTTrackHandle, this_l1track);
+    this_l1track++;
+
+    float pt = l1track_ptr->momentum().perp();
+    float eta = l1track_ptr->momentum().eta();
+    float phi = l1track_ptr->momentum().phi();
+    float z0 = l1track_ptr->z0();  //cm
+    float x0 = l1track_ptr->POCA().x();
+    float y0 = l1track_ptr->POCA().y();
+    float d0 = -x0 * sin(phi) + y0 * cos(phi);
+    float rinv = l1track_ptr->rInv();
+    float chi2rphi = l1track_ptr->chi2XYRed();
+    float chi2rz = l1track_ptr->chi2ZRed();
+    float bendchi2 = l1track_ptr->stubPtConsistency();
+    float MVA1 = l1track_ptr->trkMVA1();
+    
+    std::vector<edm::Ref<edmNew::DetSetVector<TTStub<Ref_Phase2TrackerDigi_>>, TTStub<Ref_Phase2TrackerDigi_>>>
+        stubRefs = l1track_ptr->getStubRefs();
+    int nstub = (int)stubRefs.size();
+    
+    if (chi2rz < chi2rzMax_ && MVA1 > promptMVAMin_ && pt > ptMin_ && fabs(eta) < etaMax_) {
+      if (fabs(d0) > dispD0Min_) {
+        if (MVA1 <= promptMVADispTrackMin_)
+          continue;
+      }
+      if (fabs(eta) > overlapEtaMin_ && fabs(eta) < overlapEtaMax_) {
+        if (nstub <= overlapNStubsMin_)
+          continue;
+      }
+      if (fabs(eta) > diskEtaMin_) {
+        if (fabs(d0) <= diskD0Min_)
+          continue;
+      }
+      if (fabs(eta) <= diskEtaMin_) {
+        if (fabs(d0) <= barrelD0Min_)
+          continue;
+      }
+
+      Track_Parameters track = Track_Parameters(pt,
+                                                -d0,
+                                                z0,
+                                                eta,
+                                                phi,
+                                                (1/rinv),
+                                                (this_l1track - 1),
+                                                nstub,
+                                                chi2rphi,
+                                                chi2rz,
+                                                bendchi2,
+                                                MVA1);
+
+      edm::Ptr<TrackingParticle> my_tp = MCTruthTTTrackHandle->findTrackingParticlePtr(l1track_ptr);
+      selectedTracksWithTruth.push_back(std::make_pair(track, my_tp));
+    }
+  }
+
+  //Emulation track selection loop
+  this_l1track = 0;
+  for (iterL1Track = TTTrackHandle->begin(); iterL1Track != TTTrackHandle->end(); iterL1Track++) {
+    edm::Ptr<TTTrack<Ref_Phase2TrackerDigi_>> l1track_ptr(TTTrackGTTHandle, this_l1track);
+    edm::Ptr<TTTrack<Ref_Phase2TrackerDigi_>> l1track_ref(TTTrackHandle, this_l1track);
+    this_l1track++;
+    
+    float pt = FloatPtFromBits(*l1track_ptr);
```

Comment:
What was the reason for casting things to float and needing seperate functions to extract various parameters, why can't the internal representations in L1TrackUnpacker.h?

Reply:
I was following the example of the Track Triplet Emulator which Benjamin had recommended since it was merged: https://github.com/gkaratha/cmssw/blob/ff352864a0da1cc97490466b89b6f70deeba07f2/L1Trigger/L1TTrackMatch/plugins/L1TrackTripletEmulatorProducer.cc#diff-a82ab3ec6f47ebc414fc95786550558ba75df8d73d6f91c91652c348783701df. Converting the ap_ints to floats was simpler so I could use the same functions as cut parameters as the simulation which is run in parallel with the emulation